server/balancer.go

```diff
@@ -370,3 +372,262 @@ func (r *replicaChecker) checkBestReplacement(region *RegionInfo) Operator {
+			antiCount:      r.antiCount,
+			version:        r.version,
+		})
+		status.RegionCount++
+	}
+
+	for _, rs := range l.scoreStatus {
+		sort.Sort(sort.Reverse(rs.RegionsStat))
+	}
+}
+
+func (l *balanceHotRegionScheduler) SelectSourceRegion(cluster *clusterInfo) *RegionInfo {
+	var (
+		maxWritten             uint64
+		sourceStore            uint64
+		maxHotStoreRegionCount int
+	)
+	// choose a hot store as transfer source
+	// the numbers of the hot regions in that store has higher priority than TotalWrittenBytes
+	for sid, s := range l.scoreStatus {
+		if s.RegionsStat.Len() < 2 {
+			continue
+		}
+
+		if maxHotStoreRegionCount < s.RegionsStat.Len() {
+			maxHotStoreRegionCount = s.RegionsStat.Len()
+			maxWritten = s.TotalWrittenBytes
+			sourceStore = sid
+			continue
+		}
+
+		if maxHotStoreRegionCount == s.RegionsStat.Len() && maxWritten < s.TotalWrittenBytes {
+			maxWritten = s.TotalWrittenBytes
+			sourceStore = sid
+		}
+	}
+
+	if sourceStore == 0 {
+		return nil
+	}
+
+	length := l.scoreStatus[sourceStore].RegionsStat.Len()
+	// the hottest region in the store not move
+	// radmonly pick a region from 1 .. length-1
+	// TODO: consider hot degree when pick
+	rr := l.r.Int31n(int32(length-1)) + 1
+	pickedRegionStat := l.scoreStatus[sourceStore].RegionsStat[rr]
+	if pickedRegionStat.antiCount < hotRegionAntiCount {
+		return nil
+	}
+	sourceRegion := cluster.getRegion(pickedRegionStat.RegionID)
+	if len(sourceRegion.DownPeers) != 0 || len(sourceRegion.PendingPeers) != 0 {
+		return nil
+	}
+	// use written bytes per second
+	sourceRegion.WrittenBytes = pickedRegionStat.WrittenBytes
+	l.adjustBalanceLimit(sourceStore)
+	return sourceRegion
+}
+
+func (l *balanceHotRegionScheduler) adjustBalanceLimit(storeID uint64) {
+	s := l.scoreStatus[storeID]
+	var hotRegionTotalCount float64
+	for _, m := range l.scoreStatus {
+		hotRegionTotalCount += float64(m.RegionsStat.Len())
+	}
+
+	avgRegionCount := hotRegionTotalCount / float64(len(l.scoreStatus))
+	// Multiplied by 0.75 to avoid transfer back and forth
+	limit := uint64((float64(s.RegionsStat.Len()) - avgRegionCount) * 0.75)
+	l.limit = maxUint64(1, limit)
+}
+
+func (l *balanceHotRegionScheduler) GetStatus() map[uint64]*StoreHotRegions {
+	return l.scoreStatus
+}
+
+func (l *balanceHotRegionScheduler) selectTransferLeader(sourceRegion *RegionInfo) *metapb.Peer {
+	followPeers := sourceRegion.GetFollowers()
+	storeIDs := make([]uint64, 0, len(followPeers))
+	for _, peer := range followPeers {
+		storeIDs = append(storeIDs, peer.GetStoreId())
+	}
+
+	targetStoreID := l.selectBestStore(storeIDs, sourceRegion)
+	var targetPeer *metapb.Peer
+	for _, peer := range followPeers {
+		if peer.GetStoreId() == targetStoreID {
+			targetPeer = peer
+		}
+	}
+	return targetPeer
+}
+
+func (l *balanceHotRegionScheduler) selectTransferPeer(sourceRegion *RegionInfo, cluster *clusterInfo) *metapb.Peer {
+	filter := newExcludedFilter(sourceRegion.GetStoreIds(), sourceRegion.GetStoreIds())
```

Comment:
Do we have to consider store labels when move peer?
FYI, the `balanceRegionScheduler` use `distinctScoreFilter` to make sure replicas of regions are separated as much as possible.